datastore/google/cloud/datastore/_gax.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for making API requests via GAX / gRPC."""
+
+
+import contextlib
+
+from grpc import StatusCode
+
+from google.cloud._helpers import make_insecure_stub
+from google.cloud._helpers import make_secure_stub
+from google.cloud import exceptions
+
+from google.cloud.grpc.datastore.v1 import datastore_pb2_grpc
+
+
+_GRPC_ERROR_MAPPING = {
+    StatusCode.UNKNOWN: exceptions.InternalServerError,
+    StatusCode.INVALID_ARGUMENT: exceptions.BadRequest,
+    StatusCode.DEADLINE_EXCEEDED: exceptions.GatewayTimeout,
+    StatusCode.NOT_FOUND: exceptions.NotFound,
+    StatusCode.ALREADY_EXISTS: exceptions.Conflict,
+    StatusCode.PERMISSION_DENIED: exceptions.Forbidden,
+    StatusCode.UNAUTHENTICATED: exceptions.Unauthorized,
+    StatusCode.RESOURCE_EXHAUSTED: exceptions.TooManyRequests,
+    StatusCode.FAILED_PRECONDITION: exceptions.PreconditionFailed,
+    StatusCode.ABORTED: exceptions.Conflict,
+    StatusCode.OUT_OF_RANGE: exceptions.BadRequest,
+    StatusCode.UNIMPLEMENTED: exceptions.MethodNotImplemented,
+    StatusCode.INTERNAL: exceptions.InternalServerError,
+    StatusCode.UNAVAILABLE: exceptions.ServiceUnavailable,
+    StatusCode.DATA_LOSS: exceptions.InternalServerError,
+}
+
+
+@contextlib.contextmanager
+def _grpc_catch_rendezvous():
+    """Remap gRPC exceptions that happen in context.
+
+    .. _code.proto: https://github.com/googleapis/googleapis/blob/\
+                    master/google/rpc/code.proto
+
+    Remaps gRPC exceptions to the classes defined in
+    :mod:`~google.cloud.exceptions` (according to the description
+    in `code.proto`_).
+    """
+    try:
+        yield
+    except exceptions.GrpcRendezvous as exc:
+        error_code = exc.code()
+        error_class = _GRPC_ERROR_MAPPING.get(error_code)
```

Comment:
No, this impl. predates that discussion. (The code in `_gax.py` in this PR is just being cut-and-pasted from `_http.py`.) However, it solves essentially the same problem and has the same flaws discussed there: the mapping to HTTP codes is imperfect / loses some data.